### DISCUSSION TITLE
Getting to grips with the USB Host Shield Library

### DISCUSSION THREAD

**Discussion by raichea** [marcel-licence/esp32_usb_midi#11] [Q&A] [answered]
Hi, I've got some experience with USB MIDI having developed the MistralXG project back in 2009. I'm still playing with MIDI a little and have been programming ESP devices for a few years. I'm just starting to play with the USB Host Shield 2.0 library and found your great projects. My initial task is to get familiar with the library - I'm finding the documentation a bit sparse and the examples seem very basic.

I thing i noticed was that at line 112 of usbMidiHost.ino you say, "/* this function comes from the USB Midi example of the library */" in what looks like a way to get the current USB status, but I see no similar code in any of the MIDI examples - can you provide a pointer, please?

I'm guessing you've done it all the hard way, by experimentation, so well done - I guess I'm going to have to do that too!

Cheers, Steve

**Comment by marcel-licence** [comment, marked as answer]
Hi, thank you very much for opening the discussion. Regarding line 112; I think it comes from https://github.com/felis/USB_Host_Shield_2.0/blob/master/examples/USBH_MIDI/USBH_MIDI_dump/USBH_MIDI_dump.ino line 44 ff. There is no check regarding the UsbTaskState but it does call Usb.Task and UsbMidi_Poll(). I've added a lot of output to get information about what is currently going on. Hope this project does still work. 
I moved some stuff into the library to avoid having a copy of this file in each project: https://github.com/marcel-licence/ML_SynthTools/blob/main/src/usbMidiHost.h
If it still works it would be easy to integrate in each synth project by setting up the MIDI_VIA_USB_ENABLED define.

Did you try the USB host functionality of the ESP32 S2, S3? It may work without the USB shield, but I didn't found enough time to take a look at it yet.

**Reply by raichea** [reply]
Ok, thanks. I thought you were referring to the function as a whole. No problem - as I said, well done for experimenting with it, I'll be doing much the same, it seems!

I've only recently got hold of S2/S3 devices and it took me a while to figure out how to get the serial interface working over USB, so I've not done much with them yet. I'm using PlatformIO rather than the Arduino IDE and discovered CDC_ON_BOOT by digging around. It's a bit more obvious in the Arduino IDE.
